# Практическая работа №1: Исследование алгоритмов формирования аддитивных цепочек

Выполнил студент гр. 0304 Максименко Егор, вариант 40.

## Цель работы

Формирование представления о аддитивных цепочках, выработать умение составлять и применять алгоритмы для нахождения минимальных аддитивных цепочек для заданного числа, привить навык использования систем компьютерной математики для реализации алгоритмов.

## Основные теоретические положения

$\quad$ Аддитивная цепочка - последовательность натуральных чисел от 1 до n
$$ 1 = a_0, a_1, ..., a_r = n ,$$ 
где каждый элемент равен сумме каких-либо двух предыдущих элементов
$$ a_{i} = a_{j} + a_{k}, k \leq j < i $$

$ l(n) $ - наименьшая длина аддитивной цепочки для заданного $ n \in \mathbb{N} $.

Неравенство для метода множителей: $ \quad l(mn) \leq l(m) + l(n) $

Неравенство для m-арного метода $ [m = 2^k, n = \sum_{j=0}^{k} d_{j}m^{t-j}] $: $ \quad l(n) = m - 2 + (k+1)t $

Неравенство для бинарного алгоритма "SX": $ \quad l(n) = \lambda(n) + \nu(n) - 1 $

### Свойства аддитивных цепочек
* Полагается, что все аддитивные цепочки - возрастающие:
$ 1 = a_0 < a_1 < ... < a_r = n $
* Если два числа из последовательности $\{a_i\}$ совпадают, то одно из них может быть опущено
* Пара $ (j, k), 0 \leq k < j < i $ называется шагом
* Если существует более одной пары $ (j, k) $, полагается, что $ j $ - наибольшее из возможных 

### Виды шагов 
* удвоение: $ \quad j = k = i - 1 $
* звездный шаг: $ \quad j = i - 1 $
* малый шаг: $ \quad \lambda(a_{i}) = \lambda(a_{i-1}) $

### Свойства шагов
* Первый шаг - всегда удвоение
* Удвоение - звездный шаг, но никогда не малый шаг
* За удвоением всегда следует звездный шаг
* Если $ i-й $ шаг не малый, то $ (i + 1)-й $ - либо звездный, либо малый либо и тот, и другой
* Если $ (i + 1)-й $ шаг ни звездный, ни малый, то $ i-й $ шаг должен быть малым

### Теорема
Если аддитивная цепочка включает $ d $ удвоений и $ f = r - d $ неудвоений, то 
$$ n \leq 2^{d-1}F_{f+3}, $$
где $ F_j $ - число Фибоначчи под номером $ j $.
### Следствие из теоремы
Если аддитивная цепочка содержит $ f $ удвоений и $ S $ малых шагов, то
$$ S \leq f \leq \frac{S}{1-log_2(\varphi)}, $$ 
где $ \varphi = \frac{\sqrt{5} + 1}{2} $ - золотое сечение.

### Алгоритм Брауэра
Для заданного числа $ n \in \mathbb{N} $ можно построить цепочку Брауэра с помощью рекуррентной формулы
$$ B_k(n) = 
\begin{cases} 
1, 2, 3, ..., 2^{k} - 1,\  если \  n < 2^{k}\\
B_k(q), 2q, ..., 2^kq, n,\  если \  n \geq 2^{k} и \  q = \lfloor{\frac{n}{2^k}} \rfloor
\end{cases} $$

Данная цепочка будет иметь длину
$$ l_B(n) = j(k + 1) + 2^k - 2 $$
при условии, что $ jk \leq \lambda(n) < (j + 1)k $

Длина цепочки будет минимизирована для больших $ n $, если 
$$ k = \lambda(\lambda(n)) - 2\lambda(\lambda(\lambda(n))) $$

#### Уточнения
* Задается фикированный параметр $ k $. Выполняются вычисления вспомогательных чисел:
$$ 
d = 2^k, q_1 = [\frac{n}{d}], r_1 = n \ mod \ d \Rightarrow n = q_{1}d + r_{1} \ (0 \leq r_{1} < d) \\
q_2 = [\frac{q_1}{d}], r_2 = q_1 \ mod \ d \Rightarrow q_1 = q_{2}d + r_{2}
$$
* Данная процедура выполняется до тех пор, пока $ q_s $ не станет меньше $ d $. Следовательно $ q_{s-1} = q_{s}d + r_{s} $
* Таким образом,
$$ n = 2^kq_1 + r_1 = ... = 2^k(2^k(...(2^kq_s + r_s)...) + r_2) + r_1 $$
$$ B_k(n): 1, 2, 3, ..., 2^k - 1, 2q_s, 4q_s, ..., 2^kq_s + r_s, ..., n $$

### Алгоритм Яо
* Имеет ту же вычислительную сложность, что и алгоритм Брауэра
* Раскладываем число $ n \in \mathbb{N} $ в $ 2^k-ой $ системе счисления:
$$ n = \sum_{i=0}^j a_{j}2^{ik}, \ a_j \neq 0 $$
* Введем функцию $ d(z) $:
$$ d(z) = \sum_{i: a_i = z} 2^{ik} $$

#### Ход алгоритма
* Задаем базовую последовательность: $ 1, 2, 4, ..., 2^{\lambda(n)} $
* Вычисляем значения $ d(z) $ для $ z \in \{ 1, 2, 3, ..., 2^k - 1 \} $, причем $ d(z) \neq 0 $
* Вычисляем $ zd(z) $ для всех $ z $ 
* В конечном итоге получаем
$$ n = \sum_{z = 1}^{2^k - 1} zd(z) $$

## Постановка задачи

Реализовать точные и приближённые алгоритмы нахождения минимальных аддитивных цепочек с использованием системы компьютерной математики SageMath, провести анализ алгоритмов. Полученные результаты содержательно проинтерпретировать.

## Выполнение работы
### 1. 
Вручную (т.е. **не реализовывая алгоритм** на Sage) построить последовательность вычислений бинарным методом и методом множителей для $ x^n $ для 2-3 значений 𝑛 (значение $ n \geq 30 $ выбираются студентом самостоятельно). Сравнить количество операций для каждого метода, сделать выводы.

**n = 47**

`Бинарный метод`

|   | N  | Y    | Z    |
|---|----|------|------|
| 0 | 47 | 1    | x    |
| 1 | 23 | x    | x^2  |
| 2 | 11 | x^3  | x^4  |
| 3 | 5  | x^7  | x^8  |
| 4 | 2  | x^15 | x^16 |
| 5 | 1  | x^15 | x^32 |
| 6 | 0  | x^47 | x^32 |

`Метод множителей`

47 - простое число,
$$ x^{47} = x \cdot x^{46} $$
$ 46 = 2 \cdot 23 $,
$$ x^{46} = (x^{23})^{2} $$
23 - простое число,
$$ x^{23} = x \cdot x^{22} $$ 
$ 22 = 2 \cdot 11 $,
$$ x^{22} = (x^{11})^{2} $$ 
11 - простое число,
$$ x^{11} = x \cdot x^{10}  $$
$ 10 = 2 \cdot 5 $,
$$ x^{10} = (x^{5})^{2} $$
5 - простое число,
$$ x^{5} = x \cdot x^{4} $$
$ 4 = 2 \cdot 2 $,
$$ x^{4} = (x^{2})^{2} $$

$$ x^{47} = x \cdot (x \cdot (x \cdot (x \cdot (x^{2})^{2})^{2})^{2})^{2} $$

Аддитивная цепочка: $ 1, 2, 4, 5, 10, 11, 22, 23, 46, 47 $

**n=61**

`Бинарный метод`

|   | N  | Y    | Z    |
|---|----|------|------|
| 0 | 61 | 1    | x    |
| 1 | 30 | x    | x^2  |
| 2 | 15 | x    | x^4  |
| 3 | 7  | x^5  | x^8  |
| 4 | 3  | x^13 | x^16 |
| 5 | 1  | x^29 | x^32 |
| 6 | 0  | x^61 | x^32 |

`Метод множителей`

61 - простое число,
$$ x^{61} = x \cdot x^{60} $$
$ 60 = 2 \cdot 30 $,
$$ x^{60} = (x^{30})^{2} $$
$ 30 = 2 \cdot 15 $,
$$ x^{30} = (x^{15})^{2} $$
$ 15 = 3 \cdot 5 $,
$$ x^{15} = (x^{5})^{3} $$
3 - простое число,
$$ x^{3} = x \cdot x^{2} $$
$$ => x^{15} = (x^{5})^{3} = x^{5} \cdot (x^{5})^{2} $$
5 - простое число,
$$ x^{5} = x \cdot x^{4} $$
$ 4 = 2 \cdot 2 $,
$$ x^{4} = (x^{2})^{2} $$

$$ x^{61} = x \cdot (((x \cdot (x^{2})^{2}) \cdot (x \cdot (x^{2})^{2})^{2})^{2})^{2} $$

Аддитивная цепочка: $ 1, 2, 4, 5, 10, 15, 30, 60, 61 $

**n=66**

`Бинарный метод`

|   | N  | Y    | Z    |
|---|----|------|------|
| 0 | 66 | 1    | x    |
| 1 | 33 | 1    | x^2  |
| 2 | 16 | x^2  | x^4  |
| 3 | 8  | x^2  | x^8  |
| 4 | 4  | x^2  | x^16 |
| 5 | 2  | x^2  | x^32 |
| 6 | 1  | x^2  | x^64 |
| 7 | 0  | x^66 | x^64 |

`Метод множителей`

$ 66 = 2 \cdot 33 $,
$$ x^{66} = (x^{33})^{2} $$ 
$ 33 = 3 \cdot 11 $,
$$ x^{33} = (x^{11})^{3} $$
3 - простое число,
$$ x^{3} = x \cdot x^{2} $$
$$ => x^{33} = (x^{11})^{3} = x^{11} \cdot (x^{11})^{2} $$
11 - простое число,
$$ x^{11} = x \cdot x^{10} $$
$ 10 = 2 \cdot 5 $,
$$ x^{10} = (x^{5})^{2} $$
5 - простое число,
$$ x^{5} = x \cdot x^{4} $$
$ 4 = 2 \cdot 2 $,
$$ x^{4} = (x^{2})^{2} $$

$$ x^{66} = ((x \cdot (x \cdot (x^{2})^{2})^{2}) \cdot (x \cdot (x \cdot (x^{2})^{2})^{2})^{2})^{2} $$

Аддитивная цепочка: $ 1, 2, 4, 5, 10, 11, 22, 33, 66 $

**Сравнение результатов**

| n  | Бинарный метод | Метод множителей |
|----|----------------|------------------|
| 47 | 9              | 9                |
| 61 | 9              | 8                |
| 66 | 7              | 8                |

**Вывод**

Проанализировав данные примеры (n = 47, n = 61 и n = 66), можно сделать вывод, что если в числе количество единиц в двоичной записи превышает количество нулей ($ n = 47_{10} = 101111_{2} $, $ n = 61_{10} = 111101_{2} $), то бинарный метод работает в среднем медленнее, чем метод множителей. Если же количество нулей в двоичной записи числа преобладает ($ n = 66_{10} = 1000010_{2} $), то бинарный метод работает за меньшее число операций, чем метод множителей.

### 2.
Реализовать алгоритм Брауэра (**для нечётных вариантов**) и алгоритм Яо (**для чётных вариантов**) для вычисления приближённых аддитивных цепочек для различных чисел при варьировании параметра 𝑘, сопоставить длины полученных аддитивных цепочек с минимальной аддитивной цепочкой для заданного числа. Сделать выводы.


**Реализация алгоритма Яо**

In [1]:
from sage.functions.log import logb

# Функция перевода в другую систему счисления (с основанием 2^k)
# В результате возвращает цифры числа в новой системе счисления 
# (записанные в 10-ном представлении) по возрастанию старшинства
def digits_of_converted_number(number, k):
    mask = 0
    mask_len = k
    digits = []
    for i in range(k):
        # Получение маски для извлечения цифр
        mask |= (1 << i)
    # Количество битов в записи числа
    bcount = number.nbits()
    for i in range(0, bcount, k):
        # Извлечение цифр числа в СС с основанием 2^k
        digits.append((i, (number & (mask << i)) >> i))
    return digits

# Функция для построения базовой последовательности в алгоритме Яо
def yao_base_sequence(number):
    # Вычисление значение лямбда = lb(number)
    lambda_n = floor(logb(number, 2))
    # Построение последовательности
    base_sequence = [2**i for i in range(0, lambda_n + 1)]
    return base_sequence

# Функция для подсчета значений d(z) для всех целых значений z из интервала [1; 2^k-1]
def calculate_dz_function(number, k, chain):
    # Получение всех значений z
    z_values = [i for i in range(1, 2**k)]
    # Перевод числа number в СС с основанием 2^k
    digits = digits_of_converted_number(number, k)
    
    dz_values = []
    for z in z_values:
        # Получение всех позиций цифры z в представлении числа number
        indexes = list(filter(lambda x: x[1] == z, digits))
        if indexes:
            # Расчет значения функции d(z) для данного z
            dz_values.append((z, sum(map(lambda x: 2**x[0], indexes))))
    return dz_values

# Функция для подсчета значений zd(z)
# Значение zd(z) вычисляется бинарным методом
def calculate_zdz(dz_values, chain):
    zdz = []
    
    for z, dz in dz_values:
        # Добавляем в цепочку элемент dz
        if z != 0 and dz not in chain:
            chain.append(dz)
        last_z = dz
        last_y = 0
        while z != 0:
            if z % 2 == 0:
                z //= 2
            else:
                z //= 2
                last_y += last_z
            last_z += last_z
            # Добавляем в цепочку промежуточные результаты работы бинарного алгоритма
            if last_y and last_y not in chain:
                chain.append(last_y)
            if last_z not in chain and z != 0:
                chain.append(last_z)
        # Добавляем рассчитанное значение zd(z) в массив
        zdz.append(last_y)
    return zdz
                

def yao_algorithm(number, k):
    chain = []
    # Добавляем в цепочку базовую последовательность
    chain += yao_base_sequence(number)
    # Рассчитываем значения функции d(z)
    dz_values = calculate_dz_function(number, k, chain)
    # Рассчитываем значения zd(z)
    zdz = calculate_zdz(dz_values, chain)
    result = 0
    for element in zdz:
        # Вычисляем с помощью элементов zd(z) значение number
        result += element
        if result not in chain:
            # Добавляем промежуточные результаты в цепочку
            chain.append(result)
    return chain

numbers = [5614, 3095, 1328, 1024, 3467]
# Минимальные длины аддитивных цепочек для данных чисел
# Данные были получены с помощью сервиса http://wwwhomes.uni-bielefeld.de/achim/addition_chain.html
min_lens = {5614: 16, 3095: 15, 1328: 12, 1024: 10, 3467: 15}
k_values = [i for i in range(2, 7)]
for n in numbers:
    print("Длина минимальной аддитивной цепочки для n =", n, "составляет", min_lens[n])
    for k in k_values:
        chain = yao_algorithm(n, k)
        print("n = ", n)
        print("k = ", k)
        print("Аддитивная цепочка: ")
        pretty_print(chain)
        print("Длина аддитивной цепочки: ", len(chain) - 1, end="\n\n")
    print(end="\n\n\n")


Длина минимальной аддитивной цепочки для n = 5614 составляет 16
n =  5614
k =  2
Аддитивная цепочка: 


[1,
 2,
 4,
 8,
 16,
 32,
 64,
 128,
 256,
 512,
 1024,
 2048,
 4096,
 5376,
 17,
 34,
 68,
 136,
 204,
 5410,
 5614]

Длина аддитивной цепочки:  20

n =  5614
k =  3
Аддитивная цепочка: 


[1,
 2,
 4,
 8,
 16,
 32,
 64,
 128,
 256,
 512,
 1024,
 2048,
 4096,
 40,
 6,
 192,
 448,
 5120,
 5160,
 5166,
 5614]

Длина аддитивной цепочки:  20

n =  5614
k =  4
Аддитивная цепочка: 


[1,
 2,
 4,
 8,
 16,
 32,
 64,
 128,
 256,
 512,
 1024,
 2048,
 4096,
 1280,
 17,
 34,
 68,
 102,
 136,
 238,
 5376,
 5614]

Длина аддитивной цепочки:  21

n =  5614
k =  5
Аддитивная цепочка: 


[1,
 2,
 4,
 8,
 16,
 32,
 64,
 128,
 256,
 512,
 1024,
 2048,
 4096,
 5120,
 6,
 14,
 96,
 224,
 480,
 5134,
 5614]

Длина аддитивной цепочки:  20

n =  5614
k =  6
Аддитивная цепочка: 


[1,
 2,
 4,
 8,
 16,
 32,
 64,
 128,
 256,
 512,
 1024,
 2048,
 4096,
 192,
 448,
 1472,
 6,
 14,
 46,
 5568,
 5614]

Длина аддитивной цепочки:  20




Длина минимальной аддитивной цепочки для n = 3095 составляет 15
n =  3095
k =  2
Аддитивная цепочка: 


[1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 2048, 20, 1025, 2050, 3075, 3095]

Длина аддитивной цепочки:  16

n =  3095
k =  3
Аддитивная цепочка: 


[1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 2048, 3072, 3, 7, 3088, 3095]

Длина аддитивной цепочки:  16

n =  3095
k =  4
Аддитивная цепочка: 


[1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 2048, 3, 7, 3072, 23, 3095]

Длина аддитивной цепочки:  16

n =  3095
k =  5
Аддитивная цепочка: 


[1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 2048, 3072, 3, 7, 23, 3095]

Длина аддитивной цепочки:  16

n =  3095
k =  6
Аддитивная цепочка: 


[1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 2048, 3, 7, 23, 3072, 3095]

Длина аддитивной цепочки:  16




Длина минимальной аддитивной цепочки для n = 1328 составляет 12
n =  1328
k =  2
Аддитивная цепочка: 


[1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 1280, 48, 1328]

Длина аддитивной цепочки:  13

n =  1328
k =  3
Аддитивная цепочка: 


[1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 48, 1280, 1328]

Длина аддитивной цепочки:  13

n =  1328
k =  4
Аддитивная цепочка: 


[1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 48, 1280, 1328]

Длина аддитивной цепочки:  13

n =  1328
k =  5
Аддитивная цепочка: 


[1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 288, 1312, 1328]

Длина аддитивной цепочки:  13

n =  1328
k =  6
Аддитивная цепочка: 


[1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 1280, 48, 1328]

Длина аддитивной цепочки:  13




Длина минимальной аддитивной цепочки для n = 1024 составляет 10
n =  1024
k =  2
Аддитивная цепочка: 


[1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024]

Длина аддитивной цепочки:  10

n =  1024
k =  3
Аддитивная цепочка: 


[1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024]

Длина аддитивной цепочки:  10

n =  1024
k =  4
Аддитивная цепочка: 


[1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024]

Длина аддитивной цепочки:  10

n =  1024
k =  5
Аддитивная цепочка: 


[1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024]

Длина аддитивной цепочки:  10

n =  1024
k =  6
Аддитивная цепочка: 


[1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024]

Длина аддитивной цепочки:  10




Длина минимальной аддитивной цепочки для n = 3467 составляет 15
n =  3467
k =  2
Аддитивная цепочка: 


[1,
 2,
 4,
 8,
 16,
 32,
 64,
 128,
 256,
 512,
 1024,
 2048,
 68,
 136,
 1025,
 2050,
 3075,
 392,
 3467]

Длина аддитивной цепочки:  18

n =  3467
k =  3
Аддитивная цепочка: 


[1,
 2,
 4,
 8,
 16,
 32,
 64,
 128,
 256,
 512,
 1024,
 2048,
 3,
 576,
 1152,
 2304,
 3456,
 11,
 3467]

Длина аддитивной цепочки:  18

n =  3467
k =  4
Аддитивная цепочка: 


[1,
 2,
 4,
 8,
 16,
 32,
 64,
 128,
 256,
 512,
 1024,
 2048,
 3,
 11,
 1280,
 3328,
 139,
 3467]

Длина аддитивной цепочки:  17

n =  3467
k =  5
Аддитивная цепочка: 


[1,
 2,
 4,
 8,
 16,
 32,
 64,
 128,
 256,
 512,
 1024,
 2048,
 3072,
 3,
 11,
 384,
 3083,
 3467]

Длина аддитивной цепочки:  17

n =  3467
k =  6
Аддитивная цепочка: 


[1,
 2,
 4,
 8,
 16,
 32,
 64,
 128,
 256,
 512,
 1024,
 2048,
 3,
 11,
 384,
 1408,
 3456,
 3467]

Длина аддитивной цепочки:  17






**Вывод**

Для рассматриваемых чисел (5614, 3095, 1328, 1024, 3467) алгоритм Яо при различных значениях k в диапазоне от 2 до 6 дает приблизительно одинаковые результаты. Совпадающую с минимальной длину аддитивной цепочки удалось получить для числа 1024 при всех k, максимально отдаленной по отношению к минимальной аддитивная цепочка получилась для числа 5614 при k = 4. Вероятнее всего, длина, намного большая длины минимальной цепочки, получилось из-за большого количества различных цифр в представлении числа 5614 в системе счисления с основанием $ 2^4 = 16 $

### 3.
Реализовать алгоритм дробления вектора индексов для нахождения минимальной звёздной цепочки для заданного числа. Протестировать алгоритм минимум для 5 значений 𝑛 > 1000. Указать, сколько времени потребовалось на поиск цепочки и какая цепочка получилась. Сравнить с предыдущими методами, сделать выводы.

In [2]:
from sage.functions.log import logb
import time

# Переход к следующему вектору:
# start_index - индекс в векторе (индексация с 0), который может быть изменен последним
# end_index - индекс в вектор, который может быть изменен первым
# delta - максимальное значение первого элемента вектора
def decrement_vector(vector, start_index, end_index, delta=1):
    # Начинаем уменьшение вектора с конца
    index = end_index
    # Смотрим подряд идущие элементы вектора со значением 1
    while index >= start_index and vector[index] == 1:
        index -= 1
    # Определили индекс элемента, который будем уменьшать
    # Сравниваем его со start_index
    if index < start_index:
        # Если значение всех элементов в вектора от start_index до end_index уже равно 1, то уменьшить не можем
        return False
    # Уменьшаем первый не равный 1 элемент
    vector[index] -= 1
    # Устанавливаем в максимальное значение все элементы вектора от index+1 до end_index
    for i in range(index + 1, end_index + 1):
        vector[i] = i + delta
    return True

# Построение звездной цепочки по переданному вектору
def calculate_chain_by_vector(vector):
    chain = [1]
    # Проход по вектору индексов
    for i in range(0, len(vector)):
        # Добавление в цепочку нового элемента
        chain.append(chain[-1] + chain[vector[i] - 1])
    return chain

# Построение звездной цепочки на основе уже построенной части
def chain_from_part_of_vector(part, chain_part):
    # Начало звездной цепочки передано вторым параметром
    chain = chain_part[:]
    # Дополняем звездную цепочку, воспользовавшись значениями из вектора
    for i in range(0, len(part)):
        chain.append(chain[-1] + chain[part[i] - 1])
    return chain

# Построение вектора индексов
def indexes_vector_division(number):
    # Задаем минимальное m
    m = ceil(logb(number, 2))
    while True:
        # Если число равно 2**m, то возвращаем вектор 1..m
        if 2**m == number:
            return list(range(1, m + 1))
        # Если максимальное число для данного m меньше числа number, то увеличиваем m
        if 2**m < number:
            m += 1
            continue
        # Запускаем рекурсивный алгоритм построения вектора индексов
        result = crush_part_of_vector(number, list(range(1, m+1)))
        # Если для данного m был получен результат, возвращаем его, иначе - увеличиваем m
        if result == False:
            m += 1
        else:
            return result
            
# Рекурсивный алгоритм дробления вектора индексов
# number - число, которое пытаемся получить
# part - часть вектора, которую будет разбивать на 2 (на предыдущем шаге это была меняющаяся часть)
# first_max_value - максимальное значение в векторе индексов, которое может принимать первый элемент из part
# chain - уже построенная цепочка на основе фиксированных частей из предыдущих шагов
#
# Возвращает вектор, который является частью вектора индексов для числа
# Если такой вектор найти не удалось, возвращает False
def crush_part_of_vector(number, part, first_max_value = 1, chain = [1]):
    # Определяем длину переданного вектора индексов
    m = len(part)
    # Если длина равна 1, проходим все возможные значения для элемента и строим цепочку
    if len(part) == 1:
        # Если можно уменьшить данный элемент, строим цепочку и проверяем соответствие числу number
        # Если уменьшить элемент нельзя либо последний элемент цепочки получается меньше number, возвращаем False
        while part[0] > 1:  
            a = chain_from_part_of_vector(part, chain)
            a = a[-1]
            if a < number:
                return False
            if number == a:
                return part
            part[0] -= 1
        return False
    # Разбиваем переданный вектор part на 2 части: фиксированную и меняющаюся
    q = m // 2
    # Меняем фиксированную часть, пока возможно
    can_dec_fixed_part = True
    while can_dec_fixed_part:
        # Строим цепочку
        all_chain = chain_from_part_of_vector(part, chain)
        # Берем элемент a, который является последним элементом цепочки, 
        # построенным из фиксированных частей на предыдущих шагах + фиксированной части на данном шаге
        a = all_chain[q + first_max_value - 1]
        # Вычисляем минимальное и максимальное значение числа, 
        # которое можно получить при такой фиксированной части
        a_min = a + m - q
        a_max = a * 2 ** (m - q)
        # Если число больше максимального для данной фиксированной части,
        # построить цепочку с переданным вектором не удастся
        if a_max < number:
            return False
        # Если число равняется минимальному для данной фиксированной части
        # возвращаем данную фиксированную часть + хвост из единиц
        if a_min == number:
            return part[:q] + [1 for _ in range(q + 1, m + 1)]
        # Если число равняется максимальному для данной фиксированной части
        # возвращаем полученный вектор
        if a_max == number:
            return part
        # Если число number лежит в интервале (a_min; a_max)
        # пробуем дробить меняющуюся часть для данного шага
        if a_min < number < a_max:
            crush = crush_part_of_vector(number, part[q:], first_max_value + q, all_chain[:q + first_max_value])
            # Если вектор построен, возвращаем его
            if crush:
                return part[:q] + crush
        # Если изменение меняющейся части не дало результатов, меняем фиксированную
        can_dec_fixed_part = decrement_vector(part, 0, q - 1, first_max_value)
    # Если при переданном векторе part цепочку получить не удалось, возвращаем False
    return False
        
        

numbers = [5614, 3095, 1328, 1024, 3467]

for n in numbers:
    print("n =", n)
    start_point = time.time()  
    vector = indexes_vector_division(n)
    print("indexes vector: ", vector)
    chain = calculate_chain_by_vector(vector)
    print("chain: ", chain)
    print("chain length: ", len(chain) - 1)
    end_point = time.time()
    print("time passed: ", end_point - start_point)

n = 5614
indexes vector:  [1, 2, 3, 4, 5, 6, 7, 8, 9, 9, 6, 2, 13, 14, 14, 13]
chain:  [1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 768, 800, 802, 1604, 3208, 4812, 5614]
chain length:  16
time passed:  3.784329652786255
n = 3095
indexes vector:  [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 3, 1, 13, 13, 4]
chain:  [1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 1028, 1029, 2058, 3087, 3095]
chain length:  15
time passed:  0.6773896217346191
n = 1328
indexes vector:  [1, 2, 3, 4, 5, 6, 7, 8, 5, 9, 11, 10]
chain:  [1, 2, 4, 8, 16, 32, 64, 128, 256, 272, 528, 1056, 1328]
chain length:  12
time passed:  0.0031397342681884766
n = 1024
indexes vector:  [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
chain:  [1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024]
chain length:  10
time passed:  0.0003063678741455078
n = 3467
indexes vector:  [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 8, 1, 13, 13, 4]
chain:  [1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 1152, 1153, 2306, 3459, 3467]
chain length:  15
time passed:  0.30841970443725586


Результаты вычисления цвездных цепочек для чисел 5614, 3095, 1328, 1024, 3467 представлено в таблице ниже:

|   n  |                                    Цепочка                                   | Длина цепочки | Время выполнения |
|:----:|:----------------------------------------------------------------------------:|---------------|:----------------:|
| 5614 | 1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 768, 800, 802, 1604, 3208, 4812, 5614 | 16            | 3.9401 сек       |
| 3095 | 1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 1028, 1029, 2058, 3087, 3095    | 15            | 0.7435 сек       |
| 1328 | 1, 2, 4, 8, 16, 32, 64, 128, 256, 272, 528, 1056, 1328                       | 12            | 0.0024 сек       |
| 1024 | 1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024                                  | 10            | 0.0002 сек       |
| 3467 | 1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 1152, 1153, 2306, 3459, 3467    | 15            | 0.4102 сек       |


Длины данных цепочек совпали длинам минимальных аддитивных цепочек для данных чисел. По сравнению с алгоритмом Яо аддитивные цепочки получаются намного короче. Однако время выполнения алгоритма поиска звездной цепочки в разы превышает время выполнения алгоритма Яо. Поэтому при построении аддитивной цепочки необходимо выбирать между скоростью работы и длиной аддитивной цепочки.

### 4.
Проверить гипотезу Шольца–Брауэра для всех натуральных $ 1 \leq n \leq 12 $ на алгоритме дробления вектора индексов. Результаты оформить в виде таблицы. Сделать выводы.

Гипотеза Шольца-Брауэра состоит в том, что для любого натурального числа n выполняется условие 
$$ l(2^n - 1) \leq l(n) + n - 1 $$
Уже доказано, что данная гипотеза верна для звездных цепочек, а также верна в общем случае для $ 1 \leq n \leq 64 $. \
Проверим с помощью алгоритма дробления индексов для построения звездной цепочки данную гипотезу при $ 1 \leq n \leq 12 $

In [3]:
for n in range(1, 13):
    start = time.time()
    number = 2**n - 1
    l1 = indexes_vector_division(number)
    l2 = indexes_vector_division(n)
    end = time.time()
    print("n =", n)
    print("number =", number)
    print("chain for " + str(number) + ":", calculate_chain_by_vector(l1))
    print("chain for " + str(n) + ":", calculate_chain_by_vector(l2))
    print("time passed:", end - start, "sec")
    if len(l1) <= len(l2) + n - 1:
        print("TRUE\t", len(l1), '<=', len(l2) + n - 1)
        print()
    else:
        print("FALSE")
        print(len(l1), ' > ', len(l2) + n - 1)
        print(l1)
        break

n = 1
number = 1
chain for 1: [1]
chain for 1: [1]
time passed: 0.0001285076141357422 sec
TRUE	 0 <= 0

n = 2
number = 3
chain for 3: [1, 2, 3]
chain for 2: [1, 2]
time passed: 0.00024771690368652344 sec
TRUE	 2 <= 2

n = 3
number = 7
chain for 7: [1, 2, 4, 6, 7]
chain for 3: [1, 2, 3]
time passed: 0.00035119056701660156 sec
TRUE	 4 <= 4

n = 4
number = 15
chain for 15: [1, 2, 4, 5, 10, 15]
chain for 4: [1, 2, 4]
time passed: 0.0003376007080078125 sec
TRUE	 5 <= 5

n = 5
number = 31
chain for 31: [1, 2, 4, 8, 10, 20, 30, 31]
chain for 5: [1, 2, 4, 5]
time passed: 0.003916025161743164 sec
TRUE	 7 <= 7

n = 6
number = 63
chain for 63: [1, 2, 4, 8, 16, 20, 21, 42, 63]
chain for 6: [1, 2, 4, 6]
time passed: 0.0058629512786865234 sec
TRUE	 8 <= 8

n = 7
number = 127
chain for 127: [1, 2, 4, 8, 16, 32, 40, 42, 84, 126, 127]
chain for 7: [1, 2, 4, 6, 7]
time passed: 0.0529630184173584 sec
TRUE	 10 <= 10

n = 8
number = 255
chain for 255: [1, 2, 4, 8, 16, 17, 34, 68, 85, 170, 255]
chain for 8:

Результаты проверки гипотезы Шольца-Брауэра для $ 1 \leq n \leq 12 $ представлены в таблице ниже:

|  n | l*($2^n - 1$) | l*(n) + n - 1 | Гипотеза Шольца-Брауэра |
|:--:|:-------------:|:-------------:|:-----------------------:|
|  1 |       0       |       0       |       Выполняется       |
|  2 |       2       |       2       |       Выполняется       |
|  3 |       4       |       4       |       Выполняется       |
|  4 |       5       |       5       |       Выполняется       |
|  5 |       7       |       7       |       Выполняется       |
|  6 |       8       |       8       |       Выполняется       |
|  7 |       10      |       10      |       Выполняется       |
|  8 |       10      |       10      |       Выполняется       |
|  9 |       12      |       12      |       Выполняется       |
| 10 |       13      |       13      |       Выполняется       |
| 11 |       15      |       15      |       Выполняется       |
| 12 |       15      |       15      |       Выполняется       |

Как можно видеть по таблице, гипотеза Шольца-Брауэра действительно выполняется для звездных цепочек чисел $ 2^n - 1 $ при $ 1 \leq n \leq 12 $

### 5.
Найти или предложить собственные модификации алгоритмов и привести описание модификаций. Реализовать модифицированные алгоритмы и сравнить их мощность.

Для проверки гипотезы Шольца-Брауэра для звездных цепочек при  $ n \geq 9 $ время работы алгоритма дробления индексов в изначальном виде было слишком велико (для значения $ n = 9 $ время работы составляло порядка 130 секунд, для значения $ n = 10 $ время работы алгоритма составляло порядка 25 минут). В связи с этим в работе использовался модифицированный алгоритм дробления вектора индексов.

Основная модификация заключается в том, что разбиение вектора индексов происходит не один раз за весь алгоритм, а рекурсивно для всех меняющихся частей вектора. Это работает таким образом, что если была обнаружена какая-то фиксированная часть, то строится звездная цепочка для фиксированной части, после чего происходит дробление уже меняющейся части. Это позволяет эффективно находить звездную цепочку для заданного числа, так как много неоптимальных значений меняющейся части вектора индексов не рассматриваются. Данная оптимизация позволила существенно сократить время работы алгоритма: для числа 511 ($ 2^9 - 1 $) звездная цепочка вычисляется уже за 0.25 секунд, для числа 1023 ($ 2^{10} - 1 $) - за 0.77 секунд. 

## Выводы

$\quad$В ходе выполнения работы были рассмотрены различные методы построения аддитивных цепочек для заданных чисел. Была рассмотрена длина аддитивной цепочки в зависимости от алгоритма. Были рассмотрены бинарный алгоритм построения аддитивной цепочки, метод множителей, алгоритм Яо, а также алгоритм построения звездной цепочки, основанный на разбиении вектора индексов. \
$\quad$Было проведено сравнение длин аддитивных цепочек, получаемых с помощью алгоритма Яо и алгоритма дробления вектора индексов, а также сравнени времени работы данных алгоритмов. По итогам сравнения было установлено, что алгоритм Яо в среднем дает длину цепочки больше, чем алгоритм дробления вектора индексов, однако у алгоритма дробления вектора индексов есть серьезный недостаток - время его выполнения, которое в разы больше, чем у алгоритма Яо. Несмотря на оптимизации, данный алгоритм для больших чисел будет работать намного медленнее, чем алгоритм Яо.